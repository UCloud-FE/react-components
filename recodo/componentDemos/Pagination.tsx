import React from 'react';
import { Pagination, Box } from '@ucloud-fe/react-components';

const Demo = () => {
    return (
        <Box container direction="column" spacing="lg">
            <Pagination total={100} simple />
            <Pagination total={100} showTotal showQuickJumper showSizeChanger />
        </Box>
    );
};

export default React.memo(Demo);
