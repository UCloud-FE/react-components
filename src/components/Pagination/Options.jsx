import React from 'react';
import PropTypes from 'prop-types';
import Select from 'components/Select';
import Button from 'components/Button';
import KEYCODE from 'interfaces/KeyCode';

import NumberInput from 'components/NumberInput';

class Options extends React.Component {
    static propTypes = {
        changeSize: PropTypes.func,
        quickGo: PropTypes.func,
        current: PropTypes.number,
        pageSizeOptions: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
        pageSize: PropTypes.number,
        buildOptionText: PropTypes.func,
        locale: PropTypes.object,
        size: PropTypes.string,
        allPages: PropTypes.number,
        select: PropTypes.object
    };

    static defaultProps = {
        pageSizeOptions: ['10', '20', '30', '40']
    };

    constructor(props) {
        super(props);

        this.state = {
            goInputText: ''
        };
    }

    buildOptionText = value => {
        return `${value} ${this.props.locale.itemsPerPage}`;
    };

    changeSize = value => {
        this.props.changeSize(Number(value));
    };

    handleChange = v => {
        this.setState({
            goInputText: v
        });
    };

    go = e => {
        let val = this.state.goInputText;
        if (val === '') {
            return;
        }
        val = isNaN(val) ? this.props.current : Number(val);
        if (e.keyCode === KEYCODE.ENTER || e.type === 'click') {
            this.setState({
                goInputText: ''
            });
            this.props.quickGo(val);
        }
    };

    render() {
        const props = this.props;
        const state = this.state;
        const locale = props.locale;
        const { size, allPages, select } = props;
        const prefixCls = `${props.rootPrefixCls}-options`;
        const changeSize = props.changeSize;
        const quickGo = props.quickGo;
        const goButton = props.goButton;
        const buildOptionText = props.buildOptionText || this.buildOptionText;
        let changeSelect = null;
        let goInput = null;
        let gotoButton = null;

        if (!(changeSize || quickGo)) {
            return null;
        }

        if (changeSize && Select) {
            const Option = Select.Option;
            const pageSize = props.pageSize || props.pageSizeOptions[0];
            const options = props.pageSizeOptions.map((opt, i) => (
                <Option key={i} value={opt}>
                    {buildOptionText(opt)}
                </Option>
            ));

            changeSelect = (
                <Select
                    className={`${prefixCls}-size-changer`}
                    value={pageSize.toString()}
                    onChange={this.changeSize}
                    size={size}
                    popover={{ getPopupContainer: triggerNode => triggerNode.parentNode }}
                    {...select}
                >
                    {options}
                </Select>
            );
        }

        if (quickGo) {
            if (goButton) {
                if (typeof goButton === 'boolean') {
                    gotoButton = (
                        <Button
                            onClick={this.go}
                            styleType="primary"
                            className={`${prefixCls}-gobutton`}
                            onKeyUp={this.go}
                            size={size}
                        >
                            {locale.jumpToConfirm}
                        </Button>
                    );
                } else {
                    gotoButton = (
                        <span className={`${prefixCls}-gobutton`} onClick={this.go} onKeyUp={this.go}>
                            {goButton}
                        </span>
                    );
                }
            }
            goInput = (
                <div className={`${prefixCls}-quick-jumper`}>
                    <NumberInput
                        size={size}
                        min={1}
                        max={allPages}
                        value={state.goInputText}
                        onChange={this.handleChange}
                        parser={input => input.replace(/[^\d]+/g, '')}
                        onKeyUp={this.go}
                        hideHandler
                        styleType="pagination"
                        suffix={<span className={`${prefixCls}-quick-jumper-text`}>{`/${allPages}`}</span>}
                    />
                    {gotoButton}
                </div>
            );
        }

        return (
            <li className={`${prefixCls}`}>
                {changeSelect}
                {goInput}
            </li>
        );
    }
}

export default Options;
