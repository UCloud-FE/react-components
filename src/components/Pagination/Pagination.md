### 说明

*   这是 Pagination 组件

### 演示

*   演示

```js {"codepath": "pagination.jsx"}
```

*   size

```js {"codepath": "size.jsx"}
```

*   total

```js {"codepath": "total.jsx"}
```

*   showTotal

```js {"codepath": "showTotal.jsx"}
```

*   pageSize

```js {"codepath": "pageSize.jsx"}
```

*   pageSizeOptions

```js {"codepath": "pageSizeOptions.jsx"}
```

*   showLessItems

```js {"codepath": "showLessItems.jsx"}
```

*   showPrevNextJumpers

```js {"codepath": "showPrevNextJumpers.jsx"}
```

*   showQuickJumper

```js {"codepath": "showQuickJumper.jsx"}
```

*   showSizeChanger

```js {"codepath": "showSizeChanger.jsx"}
```

*   showTitle

```js {"codepath": "showTitle.jsx"}
```

*   simple

```js {"codepath": "simple.jsx"}
```

*   uncontrolled

```js {"codepath": "uncontrolled.jsx"}
```
