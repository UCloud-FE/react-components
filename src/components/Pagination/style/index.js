import styled, { css } from 'styled-components';

import { inlineBlockWithVerticalMixin } from 'src/style';
import config from 'src/config';
import addDefaultThemeProps from 'src/components/ThemeProvider/addDefaultThemeProps';

const { prefixCls: _prefixCls } = config;

export const prefixCls = _prefixCls + '-pagination';

export const PaginationWrap = styled.ul(
    ({ theme: { designTokens: DT, fontSize, Height, HeightNumber } }) => css`
        font-size: ${fontSize};
        user-select: none;
        ${inlineBlockWithVerticalMixin};

        .${prefixCls} {
            &-item,
            &-prev,
            &-next,
            &-jump-prev,
            &-jump-next {
                outline: none;
                box-sizing: border-box;
                text-align: center;
                border-radius: 2px;
                cursor: pointer;
                padding: 0 2px;
            }
            &-item,
            &-prev,
            &-next,
            &-jump-prev,
            &-jump-next,
            &-total {
                margin-right: 5px;

                ${inlineBlockWithVerticalMixin};
                ${({ size }) => css`
                    min-width: ${Height[size]};
                    height: ${Height[size]};
                    line-height: ${HeightNumber[size] - 2}px;
                `};
            }

            &-item {
                color: ${DT.T_COLOR_TEXT_DEFAULT_LIGHT};
                border: 1px solid ${DT.T_COLOR_LINE_DEFAULT_LIGHT};
                background: ${DT.T_COLOR_BG_DEFAULT_LIGHT};
            }
            &-item:hover {
                color: ${DT.T_COLOR_TEXT_PRIMARY_DEFAULT};
                border: 1px solid ${DT.T_COLOR_LINE_PRIMARY_HOVER};
                background: ${DT.T_BUTTON_SECONDARY_COLOR_BG_DEFAULT};
            }
            &-prev,
            &-next {
                color: ${DT.T_COLOR_TEXT_DEFAULT_DARK};
                box-shadow: ${DT.T_SHADOW_BUTTON_DEFAULT};
                background: ${DT.T_BUTTON_SECONDARY_COLOR_BG_DEFAULT};
            }
            &-prev:hover,
            &-next:hover {
                color: ${DT.T_COLOR_TEXT_PRIMARY_DEFAULT};
                box-shadow: ${DT.T_SHADOW_BUTTON_HOVER};
            }

            &-jump-prev,
            &-jump-next {
                color: ${DT.T_COLOR_TEXT_DEFAULT_DARK};
            }
            &-jump-prev:hover,
            &-jump-next:hover {
                color: ${DT.T_COLOR_TEXT_PRIMARY_HOVER};
            }

            &-item-active,
            &-item-active:hover {
                border-color: ${DT.T_COLOR_LINE_PRIMARY_DEFAULT};
                color: ${DT.T_COLOR_TEXT_PRIMARY_DEFAULT};
                box-shadow: ${DT.T_SHADOW_BUTTON_HOVER};
                background: ${DT.T_BUTTON_SECONDARY_COLOR_BG_DEFAULT};
                cursor: default;
            }
            &-disabled,
            &-disabled:hover {
                color: ${DT.T_COLOR_TEXT_DISABLED};
                border: 1px solid ${DT.T_COLOR_LINE_DISABLED_DARK};
                box-shadow: none;
                background: ${DT.T_COLOR_BG_DISABLED_LIGHT};
                cursor: not-allowed;
            }

            &-prev-icon,
            &-next-icon {
                cursor: inherit;
            }

            &-options {
                &,
                &-size-changer,
                &-quick-jumper {
                    display: inline-block;
                    vertical-align: middle;
                }
                &-size-changer {
                    margin-right: 5px;
                }
                &-gobutton {
                    margin-left: 5px;
                }
            }
        }
    `
);
addDefaultThemeProps(PaginationWrap);
