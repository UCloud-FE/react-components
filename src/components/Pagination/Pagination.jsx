import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Icon from 'components/Icon';
import NumberInput from 'components/NumberInput';
import localeConsumerDecorator from 'components/LocaleProvider/localeConsumerDecorator';

import Pager from './Pager';
import Options from './Options';
import LOCALE from './locale/zh_CN';
import { PaginationWrap, prefixCls } from './style';

function noop() {}
function isInteger(value) {
    return typeof value === 'number' && isFinite(value) && Math.floor(value) === value;
}
function defaultItemRender(page, type, element) {
    if (type === 'prev') {
        return <Icon type="left" className={`${prefixCls}-prev-icon`} />;
    }
    if (type === 'next') {
        return <Icon type="right" className={`${prefixCls}-next-icon`} />;
    }
    if (type === 'jump-prev' || type === 'jump-next') {
        return <Icon type="more-three" className={`${prefixCls}-jump-icon`} />;
    }
    return element;
}
const Size = ['sm', 'md', 'lg'];

@localeConsumerDecorator({ defaultLocale: LOCALE, localeName: 'Pagination' })
class Pagination extends Component {
    static propTypes = {
        /** 当前页，受控 */
        current: PropTypes.number,
        /** 默认当前页，非受控 */
        defaultCurrent: PropTypes.number,
        /** 总数 */
        total: PropTypes.number,
        /** 每页数量，受控 */
        pageSize: PropTypes.number,
        /** 默认每页数量，非受控 */
        defaultPageSize: PropTypes.number,
        /** 修改回调 */
        onChange: PropTypes.func,
        /** 修改建议回调，如传入的total为10，当前pageSize为10，当前current却为2时会触发 */
        onAdvise: PropTypes.func,
        /** @ignore */
        hideOnSinglePage: PropTypes.bool,
        /** 显示分页数量调节 */
        showSizeChanger: PropTypes.bool,
        /** 是否显示3个 */
        showLessItems: PropTypes.bool,
        /** 分页数量变化回调 */
        onPageSizeChange: PropTypes.func,
        /** 显示更多页面跳转按钮 */
        showPrevNextJumpers: PropTypes.bool,
        /** 显示快速跳转，传入goButton显示跳转按钮 */
        showQuickJumper: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
        /** 显示按钮的title */
        showTitle: PropTypes.bool,
        /** 分页配置 */
        pageSizeOptions: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
        /** 简易分页 */
        simple: PropTypes.bool,
        /** @ignore */
        locale: PropTypes.object,
        /** @ignore */
        className: PropTypes.string,
        /** @ignore */
        itemRender: PropTypes.func,
        /** 尺寸 */
        size: PropTypes.oneOf(Size),
        /** @ignore */
        optionsProps: PropTypes.object
    };

    static defaultProps = {
        defaultCurrent: 1,
        total: 0,
        defaultPageSize: 10,
        onChange: noop,
        hideOnSinglePage: false,
        showPrevNextJumpers: true,
        showQuickJumper: false,
        showSizeChanger: false,
        showLessItems: false,
        showTitle: true,
        onPageSizeChange: noop,
        itemRender: defaultItemRender,
        size: 'md'
    };

    constructor(props) {
        super(props);

        const hasOnChange = props.onChange !== noop;
        const hasCurrent = 'current' in props;
        if (hasCurrent && !hasOnChange) {
            console.warn(
                'Warning: You provided a `current` prop to a Pagination component without an `onChange` handler. This will render a read-only component.'
            );
        }

        let current = props.defaultCurrent;
        if ('current' in props) {
            current = props.current;
        }

        let pageSize = props.defaultPageSize;
        if ('pageSize' in props) {
            pageSize = props.pageSize;
        }

        this.state = {
            current,
            currentInputValue: current,
            pageSize
        };
    }

    componentWillReceiveProps(nextProps) {
        if ('pageSize' in nextProps) {
            this.setState({
                pageSize: nextProps.pageSize
            });
        }

        const pageSize = 'pageSize' in nextProps ? nextProps.pageSize : this.state.pageSize;
        const current = 'current' in nextProps ? nextProps.current : this.state.current;

        const newCurrent = this.calculatePage(pageSize, nextProps.total);

        if ('current' in nextProps) {
            this.setState({
                current: nextProps.current,
                currentInputValue: nextProps.current
            });
        }
        if (current > newCurrent) {
            if (!('current' in nextProps)) {
                this.setState({
                    current: newCurrent,
                    currentInputValue: newCurrent
                });
            }
            const { onAdvise } = nextProps;
            if (onAdvise) {
                onAdvise(newCurrent, pageSize);
            }
        }
    }

    componentDidUpdate(prevProps, prevState) {
        // When current page change, fix focused style of prev item
        // A hacky solution of https://github.com/ant-design/ant-design/issues/8948
        if (prevState.current !== this.state.current && this.paginationNode) {
            const lastCurrentNode = this.paginationNode.querySelector(`.${prefixCls}-item-${prevState.current}`);
            if (lastCurrentNode && document.activeElement === lastCurrentNode) {
                lastCurrentNode.blur();
            }
        }
    }

    savePaginationNode = node => {
        this.paginationNode = node;
    };

    calculatePage = (p, total) => {
        total = total === undefined ? this.props.total : total;
        let pageSize = p;
        if (typeof pageSize === 'undefined') {
            pageSize = this.state.pageSize;
        }
        return Math.max(1, Math.floor((total - 1) / pageSize) + 1);
    };

    isValid = page => {
        return isInteger(page) && page >= 1 && page !== this.state.current;
    };

    changePageSize = size => {
        let current = this.state.current;
        const newCurrent = this.calculatePage(size);
        current = current > newCurrent ? newCurrent : current;
        if (typeof size === 'number') {
            if (!('pageSize' in this.props)) {
                this.setState({
                    pageSize: size
                });
            }
            if (!('current' in this.props)) {
                this.setState({
                    current,
                    currentInputValue: current
                });
            }
        }
        this.props.onPageSizeChange(current, size);
    };

    handleChange = p => {
        let page = p;
        if (this.isValid(page)) {
            if (page > this.calculatePage()) {
                page = this.calculatePage();
            }

            if (!('current' in this.props)) {
                this.setState({
                    current: page,
                    currentInputValue: page
                });
            }

            const pageSize = this.state.pageSize;
            this.props.onChange(page, pageSize);

            return page;
        }

        return this.state.current;
    };

    prev = () => {
        if (this.hasPrev()) {
            this.handleChange(this.state.current - 1);
        }
    };

    next = () => {
        if (this.hasNext()) {
            this.handleChange(this.state.current + 1);
        }
    };

    getJumpPrevPage() {
        return Math.max(1, this.state.current - (this.props.showLessItems ? 3 : 5));
    }

    getJumpNextPage() {
        return Math.min(this.calculatePage(), this.state.current + (this.props.showLessItems ? 3 : 5));
    }

    jumpPrev = () => {
        this.handleChange(this.getJumpPrevPage());
    };

    jumpNext = () => {
        this.handleChange(this.getJumpNextPage());
    };

    hasPrev = () => {
        return this.state.current > 1;
    };

    hasNext = () => {
        return this.state.current < this.calculatePage();
    };

    runIfEnter = (event, callback, ...restParams) => {
        if (event.key === 'Enter' || event.charCode === 13) {
            callback(...restParams);
        }
    };

    runIfEnterPrev = e => {
        this.runIfEnter(e, this.prev);
    };

    runIfEnterNext = e => {
        this.runIfEnter(e, this.next);
    };

    runIfEnterJumpPrev = e => {
        this.runIfEnter(e, this.jumpPrev);
    };

    runIfEnterJumpNext = e => {
        this.runIfEnter(e, this.jumpNext);
    };

    render() {
        /* eslint-disable no-unused-vars */
        const {
            current: _current,
            defaultCurrent,
            total,
            pageSize: _pageSize,
            defaultPageSize,
            onChange,
            onAdvise,
            hideOnSinglePage,
            showSizeChanger,
            showLessItems,
            onPageSizeChange,
            showPrevNextJumpers,
            showQuickJumper,
            showTitle,
            pageSizeOptions,
            simple,
            locale,
            className,
            itemRender,
            size,
            optionsProps,
            ...rest
        } = this.props;
        /* eslint-enable no-unused-vars */
        const { pageSize, current } = this.state;

        // When hideOnSinglePage is true and there is only 1 page, hide the pager
        if (hideOnSinglePage === true && total <= pageSize) {
            return null;
        }

        const allPages = this.calculatePage();
        const pagerList = [];
        let jumpPrev = null;
        let jumpNext = null;
        let firstPager = null;
        let lastPager = null;

        const goButton = showQuickJumper && showQuickJumper.goButton;
        const pageBufferSize = showLessItems ? 1 : 2;

        const prevPage = current - 1 > 0 ? current - 1 : 0;
        const nextPage = current + 1 < allPages ? current + 1 : allPages;

        const wrapClassName = classnames(prefixCls, className, { [`${prefixCls}-${size}`]: size });
        // 简易
        if (simple) {
            return (
                <ul className={classnames(wrapClassName, `${prefixCls}-simple`)} unselectable="unselectable" {...rest}>
                    <NumberInput
                        size={size}
                        min={1}
                        max={allPages}
                        styleType="pagination"
                        parser={input => input.replace(/[^\d]+/g, '')}
                        value={current}
                        onNumberChange={number => this.handleChange(number)}
                        suffix={<span className={`${prefixCls}-quick-jumper-text`}>{`/${allPages}`}</span>}
                    />
                </ul>
            );
        }

        if (allPages <= 5 + pageBufferSize * 2) {
            for (let i = 1; i <= allPages; i++) {
                const active = this.state.current === i;
                pagerList.push(
                    <Pager
                        locale={locale}
                        rootPrefixCls={prefixCls}
                        onClick={this.handleChange}
                        onKeyPress={this.runIfEnter}
                        key={i}
                        page={i}
                        active={active}
                        showTitle={showTitle}
                        itemRender={itemRender}
                    />
                );
            }
        } else {
            const prevItemTitle = showLessItems ? locale.prev3 : locale.prev5;
            const nextItemTitle = showLessItems ? locale.next3 : locale.next5;
            if (showPrevNextJumpers) {
                jumpPrev = (
                    <li
                        title={showTitle ? prevItemTitle : null}
                        key="prev"
                        onClick={this.jumpPrev}
                        tabIndex="0"
                        onKeyPress={this.runIfEnterJumpPrev}
                        className={`${prefixCls}-jump-prev`}
                    >
                        {itemRender(this.getJumpPrevPage(), 'jump-prev', <a className={`${prefixCls}-item-link`} />)}
                    </li>
                );
                jumpNext = (
                    <li
                        title={showTitle ? nextItemTitle : null}
                        key="next"
                        tabIndex="0"
                        onClick={this.jumpNext}
                        onKeyPress={this.runIfEnterJumpNext}
                        className={`${prefixCls}-jump-next`}
                    >
                        {itemRender(this.getJumpNextPage(), 'jump-next', <a className={`${prefixCls}-item-link`} />)}
                    </li>
                );
            }
            lastPager = (
                <Pager
                    locale={locale}
                    last
                    rootPrefixCls={prefixCls}
                    onClick={this.handleChange}
                    onKeyPress={this.runIfEnter}
                    key={allPages}
                    page={allPages}
                    active={false}
                    showTitle={showTitle}
                    itemRender={itemRender}
                />
            );
            firstPager = (
                <Pager
                    locale={locale}
                    rootPrefixCls={prefixCls}
                    onClick={this.handleChange}
                    onKeyPress={this.runIfEnter}
                    key={1}
                    page={1}
                    active={false}
                    showTitle={showTitle}
                    itemRender={itemRender}
                />
            );

            let left = Math.max(1, current - pageBufferSize);
            let right = Math.min(current + pageBufferSize, allPages);

            if (current - 1 <= pageBufferSize) {
                right = 1 + pageBufferSize * 2;
            }

            if (allPages - current <= pageBufferSize) {
                left = allPages - pageBufferSize * 2;
            }

            for (let i = left; i <= right; i++) {
                const active = current === i;
                pagerList.push(
                    <Pager
                        locale={locale}
                        rootPrefixCls={prefixCls}
                        onClick={this.handleChange}
                        onKeyPress={this.runIfEnter}
                        key={i}
                        page={i}
                        active={active}
                        showTitle={showTitle}
                        itemRender={itemRender}
                    />
                );
            }

            if (current - 1 >= pageBufferSize * 2 && current !== 1 + 2) {
                pagerList[0] = React.cloneElement(pagerList[0], {
                    className: `${prefixCls}-item-after-jump-prev`
                });
                pagerList.unshift(jumpPrev);
            }
            if (allPages - current >= pageBufferSize * 2 && current !== allPages - 2) {
                pagerList[pagerList.length - 1] = React.cloneElement(pagerList[pagerList.length - 1], {
                    className: `${prefixCls}-item-before-jump-next`
                });
                pagerList.push(jumpNext);
            }

            if (left !== 1 && showPrevNextJumpers) {
                pagerList.unshift(firstPager);
            }
            if (right !== allPages && showPrevNextJumpers) {
                pagerList.push(lastPager);
            }
        }

        const prevDisabled = !this.hasPrev();
        const nextDisabled = !this.hasNext();
        return (
            <PaginationWrap
                size={size}
                className={wrapClassName}
                unselectable="unselectable"
                innerRef={this.savePaginationNode}
                {...rest}
            >
                <li
                    title={showTitle ? locale.prevPage : null}
                    onClick={this.prev}
                    tabIndex={prevDisabled ? null : 0}
                    onKeyPress={this.runIfEnterPrev}
                    className={`${!prevDisabled ? '' : `${prefixCls}-disabled`} ${prefixCls}-prev`}
                    aria-disabled={prevDisabled}
                >
                    {itemRender(prevPage, 'prev', <a className={`${prefixCls}-item-link`} />)}
                </li>
                {pagerList}
                <li
                    title={showTitle ? locale.nextPage : null}
                    onClick={this.next}
                    tabIndex={nextDisabled ? null : 0}
                    onKeyPress={this.runIfEnterNext}
                    className={`${!nextDisabled ? '' : `${prefixCls}-disabled`} ${prefixCls}-next`}
                    aria-disabled={nextDisabled}
                >
                    {itemRender(nextPage, 'next', <a className={`${prefixCls}-item-link`} />)}
                </li>
                <Options
                    locale={locale}
                    rootPrefixCls={prefixCls}
                    changeSize={showSizeChanger ? this.changePageSize : null}
                    current={current}
                    pageSize={pageSize}
                    pageSizeOptions={pageSizeOptions}
                    quickGo={showQuickJumper ? this.handleChange : null}
                    goButton={goButton}
                    size={size}
                    allPages={allPages}
                    {...optionsProps}
                />
            </PaginationWrap>
        );
    }
}
Pagination.Size = Size;
export default Pagination;
