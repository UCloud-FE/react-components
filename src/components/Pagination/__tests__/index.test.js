import React from 'react';
import { mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';

import Pagination from 'src/components/Pagination';

jest.unmock('rc-trigger');
describe('Pagination', () => {
    test('click and keyboard', () => {
        const onChange = jest.fn();
        const wrapper = mount(<Pagination total={100} onChange={onChange} />);

        expect(wrapper.find('li').length).toBe(9);

        wrapper
            .find('li')
            .at(1)
            .simulate('click');
        expect(onChange).toHaveBeenCalledTimes(0);

        wrapper
            .find('li')
            .at(2)
            .simulate('click');
        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenLastCalledWith(2, 10);

        wrapper.find('.uc-fe-pagination-prev').simulate('click');
        expect(onChange).toHaveBeenCalledTimes(2);
        expect(onChange).toHaveBeenLastCalledWith(1, 10);

        wrapper.find('.uc-fe-pagination-next').simulate('click');
        expect(onChange).toHaveBeenCalledTimes(3);
        expect(onChange).toHaveBeenLastCalledWith(2, 10);

        wrapper.find('.uc-fe-pagination-jump-next').simulate('click');
        expect(onChange).toHaveBeenCalledTimes(4);
        expect(onChange).toHaveBeenLastCalledWith(7, 10);

        wrapper.find('.uc-fe-pagination-jump-prev').simulate('click');
        expect(onChange).toHaveBeenCalledTimes(5);
        expect(onChange).toHaveBeenLastCalledWith(2, 10);

        wrapper
            .find('.uc-fe-pagination-prev')
            .simulate('focus')
            .simulate('keyPress', { key: 'Enter' });
        expect(onChange).toHaveBeenCalledTimes(6);
        expect(onChange).toHaveBeenLastCalledWith(1, 10);

        wrapper
            .find('.uc-fe-pagination-next')
            .simulate('focus')
            .simulate('keyPress', { key: 'Enter' });
        expect(onChange).toHaveBeenCalledTimes(7);
        expect(onChange).toHaveBeenLastCalledWith(2, 10);

        wrapper
            .find('.uc-fe-pagination-jump-next')
            .simulate('focus')
            .simulate('keyPress', { key: 'Enter' });
        expect(onChange).toHaveBeenCalledTimes(8);
        expect(onChange).toHaveBeenLastCalledWith(7, 10);

        wrapper
            .find('.uc-fe-pagination-jump-prev')
            .simulate('focus')
            .simulate('keyPress', { key: 'Enter' });
        expect(onChange).toHaveBeenCalledTimes(9);
        expect(onChange).toHaveBeenLastCalledWith(2, 10);

        wrapper
            .find('li')
            .at(1)
            .simulate('focus')
            .simulate('keyPress', { key: 'Enter' });
        expect(onChange).toHaveBeenCalledTimes(10);
        expect(onChange).toHaveBeenLastCalledWith(1, 10);
    });

    test('readonly warn', () => {
        const warn = (global.console.warn = jest.fn());
        mount(<Pagination current={1} total={100} />);

        expect(warn).toHaveBeenCalledTimes(1);
        expect(warn).toHaveBeenLastCalledWith(
            'Warning: You provided a `current` prop to a Pagination component without an `onChange` handler. This will render a read-only component.'
        );
    });

    test('change pagesize', () => {
        const onPageSizeChange = jest.fn();
        const wrapper = mount(
            <Pagination
                pageSize={10}
                showSizeChanger
                onPageSizeChange={onPageSizeChange}
                total={100}
                optionsProps={{ select: { popover: { getPopupContainer: () => document.body } } }}
            />
        );

        expect(wrapper.find('li').length).toBe(10);
        wrapper.setProps({
            pageSize: 30
        });
        expect(wrapper.find('li').length).toBe(7);

        wrapper
            .find('div.uc-fe-pagination-options-size-changer')
            .at(0)
            .childAt(0)
            .simulate('click');
        expect(document.querySelectorAll('.uc-fe-popover').length).toBe(1);
        expect(document.querySelectorAll('.uc-fe-popover.uc-fe-popover-hidden').length).toBe(0);
        expect(document.querySelectorAll('div.uc-fe-popover>div>div>div').length).toBe(4);

        document.querySelectorAll('div.uc-fe-popover>div>div>div')[1].click();

        expect(onPageSizeChange).toHaveBeenCalledTimes(1);
        expect(onPageSizeChange).toHaveBeenLastCalledWith(1, 20);

        wrapper.unmount();
    });

    test('quick jumper', () => {
        const onChange = jest.fn();
        const wrapper = mount(<Pagination onChange={onChange} total={100} showQuickJumper={{ goButton: true }} />);

        expect(wrapper.find('input').length).toBe(1);
        wrapper.find('input').instance().value = '2';
        wrapper.find('input').simulate('change', 2);
        wrapper.find('.uc-fe-pagination-options button').simulate('click');
        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenLastCalledWith(2, 10);
    });

    test('hideOnSinglePage', () => {
        const wrapper = mount(<Pagination hideOnSinglePage total={0} />);
        expect(renderToJson(wrapper.render())).toMatchSnapshot();
    });

    test('onAdvise', () => {
        const onAdvise = jest.fn();

        const wrapper = mount(<Pagination current={2} total={100} onAdvise={onAdvise} />);

        expect(onAdvise).toHaveBeenCalledTimes(0);
        wrapper.setProps({
            current: 5,
            total: 50
        });
        expect(onAdvise).toHaveBeenCalledTimes(0);

        wrapper.setProps({
            total: 40
        });
        expect(onAdvise).toHaveBeenCalledTimes(1);
        expect(onAdvise).toHaveBeenCalledWith(4, 10);

        wrapper.setProps({
            current: 5
        });
        expect(onAdvise).toHaveBeenCalledTimes(2);
        expect(onAdvise).toHaveBeenCalledWith(4, 10);

        const onAdvise2 = jest.fn();
        const wrapper2 = mount(<Pagination defaultCurrent={5} total={100} onAdvise={onAdvise2} />);
        wrapper2.setProps({
            total: 30
        });
        expect(onAdvise2).toHaveBeenCalledTimes(1);
        expect(onAdvise2).toHaveBeenCalledWith(3, 10);
    });
});
