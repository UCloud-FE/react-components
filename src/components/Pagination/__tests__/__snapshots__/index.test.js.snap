// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination hideOnSinglePage 1`] = `null`;
